File: app/src/main/java/com/example/mobile6/ui/base/BaseDialog.java
package com.example.mobile6.ui.base;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.DialogFragment;
import androidx.viewbinding.ViewBinding;

import com.example.mobile6.R;

public abstract class BaseDialog<VB extends ViewBinding> extends DialogFragment {
    protected VB binding;
    protected Bundle args;

    protected abstract VB inflateBinding(LayoutInflater layoutInflater, ViewGroup container);

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        super.onCreateView(inflater, container, savedInstanceState);
        binding = inflateBinding(inflater, container);
        return binding.getRoot();
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setStyle(DialogFragment.STYLE_NO_TITLE, R.style.rounded_dialog_style);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        args = getArguments();
        initViews();
    }

    protected void initViews() {

    }

    protected void dismiss(String requestKey, Bundle bundle) {
        getParentFragmentManager().setFragmentResult(requestKey, bundle);
        dismiss();
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        binding = null;
    }
}

File: app/src/main/java/com/example/mobile6/ui/base/BaseDialog.kt
package com.example.mobile6.ui.base

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.DialogFragment
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.flowWithLifecycle
import androidx.lifecycle.lifecycleScope
import androidx.navigation.NavOptions
import androidx.navigation.fragment.findNavController
import androidx.viewbinding.ViewBinding
import com.example.mobile6.R
import kotlinx.coroutines.flow.launchIn
import kotlinx.coroutines.flow.onEach

abstract class BaseDialog<VB : ViewBinding> : DialogFragment() {
    private var _binding: VB? = null
    protected val binding get() = _binding!!

    protected abstract val bindingInflater: (layoutInflater: LayoutInflater, container: ViewGroup?) -> VB

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setStyle(STYLE_NO_TITLE, R.style.rounded_dialog_style);
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        _binding = bindingInflater.invoke(inflater, container)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        arguments?.let { processArguments(it) }
        initViews()
        initObservers()
    }

    abstract fun initViews()

    abstract fun initObservers()

    protected open fun processArguments(args: Bundle) {

    }

    protected fun <T> dismiss(requestKey: String, value: T) {
        findNavController().previousBackStackEntry?.savedStateHandle?.set(requestKey, value)
        dismiss()
    }

    protected fun <T> observeBackResult(requestKey: String, onResult: (T) -> Unit) {
        findNavController().currentBackStackEntry?.savedStateHandle?.getStateFlow<T?>(
            requestKey,
            null
        )
            ?.onEach {
                it?.let { value ->
                    onResult(value)
                }
            }?.flowWithLifecycle(viewLifecycleOwner.lifecycle, Lifecycle.State.STARTED)
            ?.launchIn(viewLifecycleOwner.lifecycleScope)
    }

    protected fun navigateTo(
        actionId: Int,
        args: Bundle? = null,
        navOptions: NavOptions? = null
    ) {
        findNavController().navigate(actionId, args, navOptions)
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }
}